ratchet-tree.go

```diff
@@ -0,0 +1,479 @@
+			}
+			pathNode.EncryptedPathSecret = append(pathNode.EncryptedPathSecret, ct)
+		}
+
+		dp.Nodes = append(dp.Nodes, pathNode)
+	}
+
+	t.setHashPath(from)
+
+	return dp, pathSecret
+}
+
+func (t *RatchetTree) Decap(from leafIndex, context []byte, path *DirectPath) []byte {
+	cp := copath(toNodeIndex(from), t.size())
+	if len(path.Nodes) != len(cp)+1 {
+		panic(fmt.Errorf("mls.rtn:Decap Malformed Directpath"))
+	}
+
+	dp := dirpath(toNodeIndex(from), t.size())
+
+	// leaf
+	if len(path.Nodes[0].EncryptedPathSecret) != 0 {
+		panic(fmt.Errorf("mls.rtn:Decap Malformed leaf node"))
+	}
+	t.Nodes[from].merge(&RatchetTreeNode{PublicKey: &path.Nodes[0].PublicKey})
+
+	// handle rest of the path now
+	var pathSecret []byte
+	var err error
+	haveSecret := false
+	for i := 0; i < len(cp); i++ {
+		curr := cp[i]
+		parent := parent(curr, t.size())
+		pathNode := path.Nodes[i+1]
+		if !haveSecret {
+			res := t.resolve(parent)
+			//if len(pathNode.EncryptedPathSecret) != len(res) {
+			//	panic(fmt.Errorf("mls.rtn: Malformed Ratchet Node"))
+			//}
+			for idx, v := range res {
+				if !t.Nodes[v].hasPrivate() {
+					continue
+				}
+				encryptedSecret := pathNode.EncryptedPathSecret[idx]
+				priv := t.Nodes[v].Node.PrivateKey
+				pathSecret, err = t.CipherSuite.hpke().Decrypt(*priv, []byte{}, encryptedSecret)
+				if err != nil {
+					panic(fmt.Errorf("mls:rtn: Ratchet node %v Decryption failure %v", v, err))
+				}
+				haveSecret = true
+			}
+		} else {
+			pathSecret = t.pathStep(pathSecret)
+		}
+
+		if haveSecret {
+			temp := newRatchetTreeNode(t.CipherSuite, pathSecret)
+			if !temp.PublicKey.equals(&pathNode.PublicKey) {
+				panic(fmt.Errorf("mls:rtn: incorrect public key"))
+			}
+			t.Nodes[dp[i+1]].merge(temp)
+		} else {
+			t.Nodes[dp[i+1]].merge(&RatchetTreeNode{PublicKey: &pathNode.PublicKey})
+		}
+
+	}
+	t.setHashPath(from)
+	return pathSecret
+}
+
+func (t *RatchetTree) AddLeaf(index leafIndex, key *HPKEPublicKey, credential *Credential) {
+
+	n := toNodeIndex(index)
+	sz := uint32(t.size())
+	if uint32(index) == sz {
+		if len(t.Nodes) == 0 {
+			t.Nodes = append(t.Nodes, OptionalRatchetNode{})
+		} else {
+			nw := int(n + 1)
+			for i := len(t.Nodes); i < nw; i++ {
+				t.Nodes = append(t.Nodes, OptionalRatchetNode{})
+			}
+		}
+	}
+
+	if t.Nodes[n].Node != nil {
+		panic(fmt.Errorf("mls.rtn:addLeaf: aff target already occupied"))
+	}
+
+	node := RatchetTreeNode{
+		PublicKey: key,
+		Cred:      credential,
+	}
+
+	t.Nodes[n].Node = &node
+	t.NumLeaves += 1
+
+	// update unmerged list
+	dp := dirpath(n, t.NumLeaves)
+	for _, v := range dp {
+		if leafIndex(v) == index || t.Nodes[v].Node == nil {
+			continue
+		}
+		t.Nodes[v].Node.AddUnmerged(index)
+	}
+	t.setHashPath(index)
+}
+
+func (t *RatchetTree) Merge(index leafIndex, key interface{}) {
+	curr := toNodeIndex(index)
+	if t.Nodes[curr].Node == nil {
+		panic(fmt.Errorf("mls.rtn.Merge: Cannot update a blank leaf"))
+	}
+	switch v := key.(type) {
```

Comment:
Don't do this.  Define `MergeFoo` methods for the different types.  Or just have a version that take a RatchetTreeNode and define some constructors.